### PR TITLE
FMWK-427 Qualifier builder API redesign

### PR DESCRIPTION
**Pull request by agrgr**
* QualifierBuilder: replace binName, binType, ctx, key with path
* Add `@Beta` annotation to QualifierBuilder
* Refactoring

### CODE REVIEW

File: src/main/java/org/springframework/data/aerospike/query/FilterOperation.java [hunk 1]
```diff
@@ -499,11 +496,15 @@ public Filter sIndexFilter(Map<QualifierKey, Object> qualifierMap) {
...
+            if (getCtxArr(qualifierMap) != null) pathPart = getCtxArrAsString(qualifierMap) + pathPart;
+            String path = String.join(".", getBinName(qualifierMap), pathPart);
+
             Exp[] arrElementsExp = collection.stream().map(item ->
                 Qualifier.builder()
-                    .setBinName(getBinName(qualifierMap))
+                    .setPath(getBinName(qualifierMap))
                     .setFilterOperation(FilterOperation.MAP_VAL_EQ_BY_KEY)
-                    .setKey(getKey(qualifierMap))
+                    .setPath(path)
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/query/FilterOperation.java:507]: 2 setPath's in a single QualifierBuilder?

File: src/main/java/org/springframework/data/aerospike/query/FilterOperation.java [hunk 2]
```diff
@@ -1277,7 +1281,8 @@ private static Exp processMetadataFieldInOrNot(Map<QualifierKey, Object> qualifi
         Exp[] listElementsExp = listOfLongs.stream().map(item ->
             Qualifier.metadataBuilder()
                 .setMetadataField(getMetadataField(qualifierMap))
-                .setFilterOperation(filterOperation)
+//                .setFilterOperation(filterOperation)
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/query/FilterOperation.java:1284]: Do we need this comment? if not let's remove

**Comment by roimenashe** [comment]
@agrgr So `path` replaces `binName` with optional context and nestedKey, but it uses the sindex unique DSL which we use only in Spring Data Aerospike, I'm not sure about that.

1. For me binName with context and nested key is more intuitive than a path with a DSL but that's personal, in most cases queries won't include CTX/nesting so `path` is less self explanatory than `binName`.
2. A new DSL which will be used in the new Client API design is in the works, check for `DSLPath` [here](https://aerospike.atlassian.net/wiki/spaces/CLIENTS/pages/3274702899/New+Client+API+Spec+2024#DSL,-expressions,-and-paths), it makes more sense for me to align with the new Client API DSL instead of with our sindex unique DSL which will probably be replaced with a new DSL.

I might be wrong, we can get into more details privately, lets talk about it.

**Comment by agrgr** [comment]
> we can get into more details privately, lets talk about it.

@roimenashe Yes, let's discuss. Don't see any issue with having path in most cases contain only binName, this is exactly the idea - to have the same field that controls everything, and not require from a user to figure out what is the correct CTX between binName and the last element (exlusive), etc. Moreover, the changes allowed to clean up QualifierBuilder API, so instead of many setters there are fewer which simplifies the usage.
Speaking about DSL - again, let's talk, I think we might add support for newer DSL in future, however, we already are using [singular path elements DSL](https://aerospike.atlassian.net/wiki/spaces/~963724540/pages/2954067989/Expression+DSL#Singular-path-elements) in the Indexed annotations, so it is more like a reuse of the existing API.

File: src/main/java/org/springframework/data/aerospike/query/FilterOperation.java (hunk 1 above)
Comment (agrgr) [reply src/main/java/org/springframework/data/aerospike/query/FilterOperation.java:507]: It was left after uncommenting, thanks.

File: src/main/java/org/springframework/data/aerospike/query/FilterOperation.java [hunk 3]
```diff
@@ -522,11 +522,14 @@ public Filter sIndexFilter(Map<QualifierKey, Object> qualifierMap) {
         public Exp filterExp(Map<QualifierKey, Object> qualifierMap) {
             // Convert NOT_IN to NOTEQ with logical AND as there is no direct support for NOT_IN query
             Collection<?> collection = getValueAsCollectionOrFail(qualifierMap);
+            String pathPart = getKey(qualifierMap).toString();
+            if (getCtxArr(qualifierMap) != null) pathPart = getCtxArrAsString(qualifierMap) + pathPart;
+            String path = String.join(".", getBinName(qualifierMap), pathPart);
             Exp[] arrElementsExp = collection.stream().map(item ->
                 Qualifier.builder()
-                    .setBinName(getBinName(qualifierMap))
+                    .setPath(getBinName(qualifierMap))
```
Comment (roimenashe) [inline src/main/java/org/springframework/data/aerospike/query/FilterOperation.java:530]: same here

**Review by roimenashe — APPROVED** [review]

**Review by reugn — APPROVED** [review]

**Pull request merged**